wgpu/src/texture/atlas.rs

```diff
@@ -0,0 +1,361 @@
+pub mod entry;
+
+mod allocation;
+mod allocator;
+mod layer;
+
+pub use allocation::Allocation;
+pub use entry::Entry;
+pub use layer::Layer;
+
+use allocator::Allocator;
+
+pub const SIZE: u32 = 2048;
+
+#[derive(Debug)]
+pub struct Atlas {
+    texture: wgpu::Texture,
+    texture_view: wgpu::TextureView,
+    layers: Vec<Layer>,
+}
+
+impl Atlas {
+    pub fn new(device: &wgpu::Device) -> Self {
+        let extent = wgpu::Extent3d {
+            width: SIZE,
+            height: SIZE,
+            depth: 1,
+        };
+
+        let texture = device.create_texture(&wgpu::TextureDescriptor {
+            size: extent,
+            array_layer_count: 2,
+            mip_level_count: 1,
+            sample_count: 1,
+            dimension: wgpu::TextureDimension::D2,
+            format: wgpu::TextureFormat::Bgra8UnormSrgb,
+            usage: wgpu::TextureUsage::COPY_DST
+                | wgpu::TextureUsage::COPY_SRC
+                | wgpu::TextureUsage::SAMPLED,
+        });
+
+        let texture_view = texture.create_default_view();
+
+        Atlas {
+            texture,
+            texture_view,
+            layers: vec![Layer::Empty, Layer::Empty],
```

Comment:
Apparently, there is no way to create a texture array with an `array_layer_count` of `1`. When we do this, we get a bunch of Vulkan validation errors because the created texture is a simple 2D one and the sampler is not compatible.

I am not sure if this is a limitation of the modern graphics backends or a `wgpu` design decision. We should probably bring it up in [the `wgpu` matrix channel](https://riot.im/app/#/room/#wgpu:matrix.org).

In any case, we need to start with at least 2 layers for now.

Reply:
Maybe we should decrease the atlas size then or make it configurable somehow. 4096 squared pixels times two is already 128 MB of data if I calculated that right.

Reply:
It's 2048 currently.

Reply:
In the long run, `wgpu` should offer a way to check available VRAM so we can choose a proper value at runtime.

Reply:
2048 is fine, I think.